### PR TITLE
ResumeDialog is not called after resumption with adaptivedialog 

### LINKED ISSUE #5401: Skill Dialog does not reprompt properly after being interrupted by another Skill Dialog
## Version
4.12.1

## Describe the bug
Give a clear and concise description of what the bug is.

I have a bot in composer with two skills connected. When I interrupt Skill A in a waiting state with Skill B, I expect Skill A to reprompt once Skill B completes, but it does not. The next message is taken as a response to Skill A's waiting dialog, so it seems like an issue with the reprompting mechanism.

### PR DESCRIPTION
**Pull request by tomlm**
Fixes #5401

## Description
* Updated AdaptiveDialog to store the interruption status in TurnPath.Interrupted so other dialogs can check it and decide if they need to reprompt or continue.
* Updated InputDialog and SkillDialog to check on interruption status to decide if it is going to reprompt or continue. 
* Update SkillConversationFactory to always return a unique (and different) ID every time CreateSkillConversationIdAsync (this should also address https://github.com/microsoft/botframework-sdk/issues/6248

### CODE REVIEW

File: libraries/Microsoft.Bot.Builder/Skills/SkillConversationIdFactory.cs [hunk 1]
```diff
@@ -47,13 +47,7 @@ public SkillConversationIdFactory(IStorage storage)
...
             var conversationReference = options.Activity.GetConversationReference();
 
-            string skillConversationId = string.Join(
-                "-",
-                options.FromBotId,
-                options.BotFrameworkSkill.AppId,
-                conversationReference.Conversation.Id,
-                conversationReference.ChannelId,
-                "skillconvo");
+            var skillConversationId = Guid.NewGuid().ToString();
```
Comment (Stevenic) [inline libraries/Microsoft.Bot.Builder/Skills/SkillConversationIdFactory.cs:50]: Much simpler :)

File: libraries/Microsoft.Bot.Builder.Dialogs/SkillDialog.cs [hunk 2]
```diff
@@ -91,6 +91,15 @@ public override async Task<DialogTurnResult> BeginDialogAsync(DialogContext dc,
         /// return value.</remarks>
         public override async Task<DialogTurnResult> ContinueDialogAsync(DialogContext dc, CancellationToken cancellationToken = default)
         {
+            // with adaptive dialogs, ResumeDialog is not called directly.  Instead the Interrupted flag is set, which
+            // acts as the signal to the SkillDialog to resume the skill.
+            if (dc.State.TryGetValue<bool>(TurnPath.Interrupted, out bool interrupted) && interrupted)
+            {
+                // resume dialog execution
+                dc.State.SetValue(TurnPath.Interrupted, false);
```
Comment (Stevenic) [inline libraries/Microsoft.Bot.Builder.Dialogs/SkillDialog.cs:99]: >dc.State.SetValue(TurnPath.Interrupted, false); [](start = 16, length = 47)

I don't know that you need to do this but shouldn't hurt anything.

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/AdaptiveDialog.cs [hunk 3]
```diff
@@ -653,22 +653,35 @@ protected async Task<DialogTurnResult> ContinueActionsAsync(DialogContext dc, ob
             // from the stack and we want to detect this so we can stop processing actions.
             var instanceId = GetUniqueInstanceId(actionContext);
 
+            // Initialize local interruption detection
+            // - Any steps containing a dialog stack after the first step indicates the action was interrupted. We
+            //   want to force a re-prompt and then end the turn when we encounter an interrupted step.
+            var interrupted = false;
```
Comment (Stevenic) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/AdaptiveDialog.cs:659]: >var interrupted = false; [](start = 12, length = 24)

This code is absolutely brilliant :)

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Input/InputDialog.cs [hunk 4]
```diff
@@ -230,16 +230,18 @@ public override async Task<DialogTurnResult> BeginDialogAsync(DialogContext dc,
         public override async Task<DialogTurnResult> ContinueDialogAsync(DialogContext dc, CancellationToken cancellationToken = default(CancellationToken))
         {
             var activity = dc.Context.Activity;
-            if (activity.Type != ActivityTypes.Message)
+
+            // Interrupted dialogs reprompt so we can ignore the incoming activity. 
+            var interrupted = dc.State.GetValue<bool>(TurnPath.Interrupted, () => false);
+            if (!interrupted && activity.Type != ActivityTypes.Message)
```
Comment (Stevenic) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Input/InputDialog.cs:236]: >if (!interrupted && activity.Type != ActivityTypes.Message) [](start = 12, length = 59)

I see the issue you were having with EoC... This makes sense...

**Review by Stevenic — APPROVED** [review]
:shipit:

File: tests/Microsoft.Bot.Builder.Tests/Skills/SkillConversationIdFactoryTests.cs [hunk 5]
```diff
@@ -53,6 +53,36 @@ public async Task SkillConversationIdFactoryHappyPath()
             Assert.Null(deletedConversationReference);
         }
 
+        [Fact]
+        public async Task IdIsUniqueEachTime()
```
Comment (johnataylor) [inline tests/Microsoft.Bot.Builder.Tests/Skills/SkillConversationIdFactoryTests.cs:57]: should this test be called IdIsUniqueTwiceInARow?

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/AdaptiveDialog.cs [hunk 6]
```diff
@@ -674,7 +674,7 @@ protected async Task<DialogTurnResult> ContinueActionsAsync(DialogContext dc, ob
                 else
                 {
                     // Set interrupted flag
-                    if (interrupted && !actionDC.State.ContainsKey(TurnPath.Interrupted))
+                    if (interrupted && !actionDC.State.TryGetValue(TurnPath.Interrupted, out _))
```
Comment (johnataylor) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/AdaptiveDialog.cs:677]: The fact that there is different semantics between these lines is disturbing. Note ContainsKey and TryGetValue are both part of the IDictionary contract they shouldn't have such differing semantics.

File: libraries/Microsoft.Bot.Builder/Skills/SkillConversationIdFactory.cs (hunk 1 above)
Comment (johnataylor) [inline libraries/Microsoft.Bot.Builder/Skills/SkillConversationIdFactory.cs:50]: much nicer

**Review by johnataylor — APPROVED** [review]
Couple of comments, however, I don't think they are actionable right now. Its _very_ hard to tell what the semantics of this loop are. But the interruption flag doesn't seem to have broken the non-interruption path at least. I think the conversation id factory is clearer now so that part is good. Approving but it would good to understand the testing.

**Pull request merged**